### PR TITLE
adding jscs and jshint to the pretest, jsdocs updates

### PR DESCRIPTION
**Pull request by clarkbw**
Creating minimal jshint and jscs configs that required mostly doc updates and little code reformatting.
Ignoring the payload module as there aren’t tests for the file.
In the gcm provider started using the ‘truthy’ test instead of the `!= null` test and added a few extra test parameters to ensure it works as advertised.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,3 +3,7 @@ node_js:
   - "0.12"
   - "0.11"
   - "iojs"
+
+before_script:
+    - npm install -g jshint jscs
+    - npm install
```
Comment (bajtos) [inline .travis.yml]: Please add `jshint` and `jscs` to `devDependencies` in package.json and revert back the changes in `.travis.yml`.

That's the only right way how to ensure that `npm install && npm test` works on all machines.

File: lib/push-manager.js [hunk 2]
```diff
@@ -20,10 +20,11 @@ module.exports = PushManager;
 
 /**
  * The PushManager class.
- *
+ * See the {@link https://github.com/tcs-de/nodecache#options|node-cache options}
+ * for more information on the settings parameters.
  * @options {Object} settings The push settings.
  * @property {Number} ttlInSeconds Time-to-live, in seconds.
- * @property {Number} checkPeriodInSeconds
+ * @property {Number} checkPeriodInSeconds A nuber in seconds for the automatic delete check interval
```
Comment (bajtos) [inline lib/push-manager.js]: typo, should be "number"?

File: lib/push-manager.js [hunk 3]
```diff
@@ -136,13 +138,19 @@ PushManager.prototype.configureProvider = function (deviceType, pushSettings) {
...
 PushManager.prototype.configureApplication = function (appId, deviceType, cb) {
   assert.ok(cb, 'callback should be defined');
   var self = this;
   var msg;
 
+  /**
+   * @callback PushManager~configureApplicationCallback
+   * @param {Error|null} err The error object
+   * @param {Application} application The Application matched to the the appId
+   */
```
Comment (bajtos) [inline lib/push-manager.js]: We use a different format for describing callback arguments, see [here](https://github.com/strongloop/loopback/blob/553a3b208386a0e946cccfcfdba1d83eae8a6f15/lib/persisted-model.js#L85-L97). I know our format is non-standard, but it is required for http://apidocs.strongloop.com/

Could you please remove this comment and describe the callback argument in the method doc block instead?

File: models/installation.js [hunk 4]
```diff
@@ -32,12 +38,10 @@ module.exports = function(Installation) {
...
   /**
    * Find installations by application id/version
-   * @param {String} deviceType The device type
+   * @param {String} deviceType The type of device (android, ios)
    * @param {String} appId The application id
    * @param {String} [appVersion] The application version
-   * @callback {Function} cb The callback function
-   * @param {Error|String} err The error object
-   * @param {Installation[]} installations The selected installations
+   * @param {Installation~findByCallback} cb The callback function passed to find()
```
Comment (bajtos) [inline models/installation.js]: Same comment as above. Either use the convention expected by apidoc.strongloop.com, or else simplify to `@param {function(Error=,Installation[])} cb ...`

File: package.json [hunk 5]
```diff
@@ -19,7 +20,7 @@
     "debug": "^2.2.0",
     "lodash": "^3.9.3",
     "mpns": "^2.1.0",
-    "node-cache": "^2.2.0",
+    "node-cache": "^2.1.1",
```
Comment (bajtos) [inline package.json]: This look suspiciously to me, why are you downgrading the minimal required version of node-cache?

File: test/gcm.provider.test.js [hunk 6]
```diff
@@ -88,7 +88,7 @@ describe('GCM provider', function() {
 
   describe('for multiple device tokens', function() {
     it('sends Notification as a GCM message', function(done) {
-      var notification = aNotification({ aKey: 'a-value' });
+      var notification = aNotification({ aKey: 'a-value', aNull: null, message: undefined });
```
Comment (bajtos) [inline test/gcm.provider.test.js]: This looks like an unrelated change to me, why is it necessary?

Comment (bajtos) [inline test/gcm.provider.test.js]: (How is this related to coding style cleanup?)

**Comment by bajtos** [comment]
Thank you for the pull request, @clarkbw . It will be great to have an automated style check in place :+1: 

See the comments above for things to fix/improve before this patch can be landed.

File: package.json (hunk 5 above)
Comment (clarkbw) [inline package.json]: Should have commented that better, see the [release history](https://www.npmjs.com/package/node-cache#release-history) for node-cache.  The 2.2.0 version was revoked and is no longer available.  I did a quick pass over the 2.1.1 for compat but I should have spent more time reviewing it.

File: test/gcm.provider.test.js (hunk 6 above)
Comment (clarkbw) [inline test/gcm.provider.test.js]: Should have added better notes for this as well.  As part of the code style changes I changed the logic test here: https://github.com/clarkbw/loopback-component-push/commit/0f571dbd6c3c8701a90b9ef4813425c2f4094476#diff-8ba8b7926d5e215abcfc9cff520946b2L71 and even though its obvious I wanted to have the test show that non-truthy values were still ignored.

File: package.json (hunk 5 above)
Comment (clarkbw) [inline package.json]: Let me know what you'd like done here.

File: .travis.yml (hunk 1 above)
Comment (clarkbw) [inline .travis.yml]: Good point.  Fixed.

File: lib/push-manager.js (hunk 2 above)
Comment (clarkbw) [inline lib/push-manager.js]: Good catch!

File: models/installation.js (hunk 4 above)
Comment (clarkbw) [inline models/installation.js]: I'll go with this shorthand format because it allows jscs to pass as configured where it checks the param names to be valid.

**Comment by clarkbw** [comment]
Thanks for the review @bajtos! I added comments inline and pushed a new commit with fixes to the issues without lingering questions.  I'll take a longer look through the node-cache version change later tonight to see if there were changes of consequence.

File: test/gcm.provider.test.js (hunk 6 above)
Comment (bajtos) [inline test/gcm.provider.test.js]: I see, I agree it's a good idea to have that behaviour tested. Please create a new test with a name that makes the intent clear, e.g. "it ignores non-truthy values" instead of modifying an unrelated test.

File: lib/providers/gcm.js [hunk 7]
```diff
@@ -68,7 +68,7 @@ GcmProvider.prototype._createMessage = function(notification) {
   });
 
   Object.keys(notification).forEach(function (key) {
-    if (notification[key] != null) {
+    if (notification[key]) {
```
Comment (bajtos) [inline lib/providers/gcm.js]: I think this is actually a breaking change, as `{ customKey: false }` is ignored now, which may be a problem. Could you please rework the implementation to ignore only `null` and `undefined`?

File: package.json (hunk 5 above)
Comment (bajtos) [inline package.json]: >  The 2.2.0 version was revoked and is no longer available.

Ah, I see, it's fine to downgrade in that case. Could you please make the downgrade in a standalone commit and describe the reasons for downgrade in the commit message?

**Comment by bajtos** [comment]
>  I'll take a longer look through the node-cache version change later tonight to see if there were changes of consequence.

No need for that, I am ok to downgrade, see my inline comment above.

File: package.json (hunk 5 above)
Comment (clarkbw) [inline package.json]: created #91 to handle that

File: lib/providers/gcm.js (hunk 7 above)
Comment (clarkbw) [inline lib/providers/gcm.js]: Good catch, I'll create a specific simple test as mentioned above to check that `null` and `undefined` are ignored while other values aren't.

**Comment by clarkbw** [comment]
@bajtos let me know if there's anything else you'd like to see.

**Comment by bajtos** [comment]
Landed, thank you for the contribution :+1:
